Initial implementation of globe camera controller

wonder-sk opened the pull request:
This is a new navigation mode for QgsCameraController which is meant to work just like the terrain-based navigation mode, but for 3D scenes with globe.

Initially, the new navigation mode can handle:
- zooming in/out with mouse wheel
- orbit around globe when dragging with left mouse button pressed
- left/right/up/down keys to orbit around globe
- shift + left/right keys to change heading angle
- shift + up/down keys to change pitch angle
- page up/page down keys to increase/decrease elevation

For camera pose, we use QgsCameraPose as for "flat" scenes, but the QCamera is positioned/rotated slightly differently - the main change is that in case of globe, we use ECEF coordinates of the camera pose's center point, convert them to lat/lon coordinates in order to set up initial rotation of the camera so that it is perpendicular to the tangent plane at the given ECEF coordinate.

These are only under the hood additions, there is currently no way to select this navigation mode in GUI.

A quick demo in 3D sandbox (the globe mesh implementation is not included - that will go to another PR):

https://github.com/user-attachments/assets/5f9b20fe-169d-4ec4-ba4b-7ac54f90d89b

wonder-sk committed:
Initial implementation of globe camera controller

This is a new navigation mode for QgsCameraController which is meant to work
just like the terrain-based navigation mode, but for 3D scenes with globe.

Initially, the new navigation mode can handle:
- zooming in/out with mouse wheel
- orbit around globe when dragging with left mouse button pressed
- left/right/up/down keys to orbit around globe
- shift + left/right keys to change heading angle
- shift + up/down keys to change pitch angle
- page up/page down keys to increase/decrease elevation

For camera pose, we use QgsCameraPose as for "flat" scenes, but the QCamera
is positioned/rotated slightly differently - the main change is that in case
of globe, we use ECEF coordinates of the camera pose's center point, convert
them to lat/lon coordinates in order to set up initial rotation of the camera
so that it is perpendicular to the tangent plane at the given ECEF coordinate.

uclaros commented on src/3d/qgscameracontroller.h:
```diff
@@ -242,6 +243,39 @@ class _3D_EXPORT QgsCameraController : public QObject
+     * and longitude (in degrees) and distance from that point on ellipsoid
+     * \since QGIS 3.44
+     */
+    void resetGlobe( float distance, double lat = 0, double lon = 0 );
```
What would you think about changing the parameters' order to use `lon` first, `lat` second?
Matching the x-y order feels more natural to me.

wonder-sk replied:
i'm easy... I was thinking I should probably switch from lat/lon min/max to QgsRectangle and from lat/lon to QgsPointXY - to avoid any confusion

nyalldawson commented on src/3d/qgscameracontroller.cpp:
```diff
@@ -304,11 +311,70 @@ double QgsCameraController::sampleDepthBuffer( int px, int py )
+    {
+      QgsVector3D viewCenter = mCameraPose.centerPoint() + mOrigin;
+
+      QgsVector3D viewCenterLatLon = mGlobeCrsToLatLon.transform( viewCenter );
```
catch QgsCsException

nyalldawson commented on src/3d/qgscameracontroller.cpp:
```diff
@@ -741,6 +881,50 @@ void QgsCameraController::onKeyPressedTerrainNavigation( Qt3DInput::QKeyEvent *e
 
+void QgsCameraController::onKeyPressedGlobeTerrainNavigation( Qt3DInput::QKeyEvent *event )
+{
+  const float moveFactor = 0.000001f; // multiplied by distance to get angle
+  const float zoomFactor = 0.9f;
```
```suggestion
  constexpr float MOVE_FACTOR = 0.000001f; // multiplied by distance to get angle
  constexpr float ZOOM_FACTOR = 0.9f;
```

nyalldawson commented on src/3d/qgscameracontroller.cpp:
```diff
@@ -741,6 +881,50 @@ void QgsCameraController::onKeyPressedTerrainNavigation( Qt3DInput::QKeyEvent *e
+  const float zoomFactor = 0.9f;
+
+  const bool hasShift = ( event->modifiers() & Qt::ShiftModifier );
+
```
How about ctrl modifier = 1/10 movement factor, like we use elsewhere?

wonder-sk replied:
yeah, that should come later, with a couple of other improvements to bring it on par with the current terrain-based navigation (e.g. zooming/rotation with pivot point away from window center) - but I'd like to keep this PR minimal in scope

wonder-sk commented on src/3d/qgscameracontroller.cpp:
```diff
@@ -59,6 +59,12 @@ QgsCameraController::QgsCameraController( Qgs3DMapScene *scene )
+  if ( mScene->mapSettings()->isGlobe() )
+  {
+    // TODO: is it okay to transform to EPSG:4326 if the geocentric CRS is based on other datum?
+    mGlobeCrsToLatLon = QgsCoordinateTransform( mScene->mapSettings()->crs(), QgsCoordinateReferenceSystem( "EPSG:4326" ), mScene->mapSettings()->transformContext() );
```
Unfortunately that returns invalid crs (tried with `QgsCoordinateReferenceSystem("EPSG:4978").toGeographicCrs()`)

I guess toGeographicCrs() only works if it is a projected CRS... and I wonder how would we make sure the axis order is the one we want (lon-lat and not lat-lon)

nyalldawson replied:
Should be fixed by https://github.com/qgis/QGIS/pull/61074

wonder-sk commented on src/3d/qgscameracontroller.cpp:
```diff
@@ -741,6 +881,50 @@ void QgsCameraController::onKeyPressedTerrainNavigation( Qt3DInput::QKeyEvent *e
 
+void QgsCameraController::onKeyPressedGlobeTerrainNavigation( Qt3DInput::QKeyEvent *event )
+{
+  const float moveFactor = 0.000001f; // multiplied by distance to get angle
```
distance from center point controls how big the step will be in the end... if the center point is away from the globe surface, that should not be a problem (assuming that the center point is set up reasonably well that it is near some actual data)

wonder-sk committed:
Add 3D scene mode enum

wonder-sk committed:
Review from Stefanos and Nyall

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wonder-sk committed:
fix tests for 4978 rendering

updated the expected images because the terrain based
camera controller no longer applies to globe views

wonder-sk committed:
add test image for qt6

wonder-sk committed:
make clang-tidy happy

wonder-sk committed:
fix second 4978 control image

wonder-sk commented:
Thanks for the reviews, this should be now ready to go...
